arrow-arith/src/temporal.rs

```diff
@@ -19,105 +19,352 @@
+    let map_func = get_date_time_tz_part_extract_fn(part);
+    array.unary_opt(|d| {
+        timestamp_ms_to_datetime(d)
+            .map(|c| Utc.from_utc_datetime(&c).with_timezone(&tz))
+            .map(map_func)
+    })
+}
+
+fn date_part_timestamp_us(
+    array: &PrimitiveArray<TimestampMicrosecondType>,
+    part: DatePart,
+) -> Int32Array {
+    let map_func = get_naive_date_time_part_extract_fn(part);
+    array.unary_opt(|d| timestamp_us_to_datetime(d).map(map_func))
+}
+
+fn date_part_timestamp_us_tz(
+    array: &PrimitiveArray<TimestampMicrosecondType>,
+    part: DatePart,
+    tz: Tz,
+) -> Int32Array {
+    let map_func = get_date_time_tz_part_extract_fn(part);
+    array.unary_opt(|d| {
+        timestamp_us_to_datetime(d)
+            .map(|c| Utc.from_utc_datetime(&c).with_timezone(&tz))
+            .map(map_func)
+    })
+}
+
+fn date_part_timestamp_ns(
+    array: &PrimitiveArray<TimestampNanosecondType>,
+    part: DatePart,
+) -> Int32Array {
+    let map_func = get_naive_date_time_part_extract_fn(part);
+    array.unary_opt(|d| timestamp_ns_to_datetime(d).map(map_func))
+}
+
+fn date_part_timestamp_ns_tz(
+    array: &PrimitiveArray<TimestampNanosecondType>,
+    part: DatePart,
+    tz: Tz,
+) -> Int32Array {
+    let map_func = get_date_time_tz_part_extract_fn(part);
+    array.unary_opt(|d| {
+        timestamp_ns_to_datetime(d)
+            .map(|c| Utc.from_utc_datetime(&c).with_timezone(&tz))
+            .map(map_func)
+    })
+}
+
+/// Given array, return new array with the extracted [`DatePart`].
+///
+/// Returns an [`Int32Array`] unless input was a dictionary type, in which case returns
+/// the dictionary but with this function applied onto its values.
+///
+/// Returns error if attempting to extract date part from unsupported type (i.e. non-date/timestamp types).
+pub fn date_part(array: &dyn Array, part: DatePart) -> Result<ArrayRef, ArrowError> {
+    downcast_primitive_array!(
+        array => {
+            let array = primitive_array_date_part(array, part)?;
+            let array = Arc::new(array) as ArrayRef;
+            Ok(array)
+        }
+        DataType::Dictionary(_, _) => {
+            let array = array.as_any_dictionary();
+            let values = date_part(array.values(), part)?;
+            let new_array = array.with_values(Arc::new(values) as ArrayRef);
+            Ok(new_array)
+        }
+        t => return_compute_error_with!(format!("{part} does not support"), t),
+    )
+}
+
+/// Dispatch to specialized function depending on the array type. Since we don't need to consider
+/// dictionary arrays here, can strictly control return type to be [`Int32Array`].
+fn primitive_array_date_part<T: ArrowPrimitiveType>(
+    array: &PrimitiveArray<T>,
+    part: DatePart,
+) -> Result<Int32Array, ArrowError> {
+    match array.data_type() {
+        DataType::Date32 => {
+            let array = downcast_array::<Date32Array>(array);
+            Ok(date_part_date32(&array, part))
```

Comment:
```suggestion
            Ok(date_part_date32(array.as_primitive::<Date32Type>(), part))
```

I prefer the `AsArray` formulation